lib/PreviewProvider.js

```diff
@@ -0,0 +1,226 @@
+    }
+
+    // remove extra slashes, and instances of '/../' and '/./' then construct back a safe pathname
+    if (newURL.pathname) {
+      let safePathItems = newURL.pathname.split("/").filter(item => item.replace(/\.{1,}/, ""));
+      let safePath = "/";
+      newURL.pathname = safePath + safePathItems.join("/");
+    }
+
+    // if the url contains sensitive information, remove it
+    if (newURL.username) {
+      newURL.username = "";
+    }
+    if (newURL.password) {
+      newURL.password = "";
+    }
+
+    newURL.hash = "";
+    return newURL.toString();
+  },
+
+  /**
+    * Create a key based on a URL in order to dedupe sites
+    */
+  _createCacheKey(url) {
+    url = new URL(url);
+    let key = url.host.replace(/www\.?/, "");
+    key = key + url.pathname;
+    return key.toString();
+  },
+
+  /**
+    * Canonicalize urls by sanitizing them, then deduping them
+    */
+  _uniqueLinks(links) {
+    let dedupedSites = new Map();
+    links.forEach(site => {
+      site.sanitizedURL = this._sanitizeURL(site.url);
+      site.cacheKey = this._createCacheKey(site.url);
+      if (!dedupedSites.has(site.cacheKey)) {
+        dedupedSites.set(site.cacheKey, site);
+      }
+    });
+    return Array.from(dedupedSites.values());
+  },
+
+  /**
+    * Filter cached links out, and store their embedly data
+    */
+  getCachedLinks(links) {
+    let cachedLinks = [];
+    let sites = this._uniqueLinks(links.filter(this._URLFilter(URL_FILTERS)));
+    sites.forEach(link => {
+      if (link && ss.storage.embedlyData[link.cacheKey] && link.cacheKey === ss.storage.embedlyData[link.cacheKey].cacheKey) {
+        cachedLinks.push(ss.storage.embedlyData[link.cacheKey]);
+      }
+    });
+
+    return cachedLinks;
+  },
+
+  /**
+   * Filter out new links and request them from embedly
+   */
+  saveNewLinks: Task.async(function*(links) {
+    let sites = this._uniqueLinks(links.filter(this._URLFilter(URL_FILTERS)));
+    let newLinks = sites.filter(link => link && !ss.storage.embedlyData[link.cacheKey]);
+    let requestQueue = [];
+    while (newLinks.length !== 0) {
+      // we have some new links we need to fetch the embedly data for, put them on the queue
+      requestQueue.push(newLinks.splice(0, EMBEDLY_PROXY_MAX_LINKS));
+    }
+    // for each bundle of 25 links, create a new request to embedly
+    requestQueue.forEach(requestBundle => {
+      this._fetchAndCache(requestBundle);
+    });
+  }),
+
+  /**
+   * Makes the necessary requests to embedly to get data and store it in the cache
+   */
+  _fetchAndCache: Task.async(function*(newLinks) {
+    let json = null;
+    // extract only the sanitized link urls to send to embedly
+    let linkURLs = newLinks.map(link => link.sanitizedURL);
+    try {
+      let response = yield fetch(EMBEDLY_ENDPOINT, {
+        method: "POST",
+        body: JSON.stringify({urls: linkURLs}),
+        headers: {"Content-Type": "application/json"}
+      });
+      if (response.ok) {
```

Comment:
So I don't know how the client wants to handle it.  A failure in this case means that there will be no data returned at all.  The proxy does not return partial results (yet).  Soon it will attempt to do result level validation and then it may throw out a result.  So later we will have to amend the client to do things like compare which URLs it sent with which it got back and possibly retry or something, but I think that can come later, for now that behaviour is poorly defined so let's just leave it out of this PR and handle it in a separate ticket.  In fact I will file that ticket.